R/teal_lockfile.R

```diff
@@ -0,0 +1,119 @@
+#'
+#' This function is invoked during [teal::init] to create `renv`-compatible lockfile for use within the application.
+#'
+#' The function leverages [renv::snapshot()], which offers multiple methods for lockfile creation.
+#'
+#' - User-specified:
+#'     - **Pre-computed lockfile**: Users can provide their own pre-computed lockfile by specifying the path via
+#'     `options(teal.renv.lockfile = "")`. In this case, this function is bypassed.
+#' - Automatically computed:
+#'     - **Working directory lockfile**: If `options("teal.renv.lockfile")` is empty, `teal` will, by default, create an
+#'      `implicit` type lockfile that uses `renv::dependencies()` to detect all R packages in the current project's
+#'      working directory.
+#'     - **`DESCRIPTION`-based lockfile**: To generate a lockfile based on a `DESCRIPTION` file in your working
+#'     directory, set `renv::settings$snapshot.type("explicit")`. The naming convention for `type` follows
+#'     `renv::snapshot()`. For the `"explicit"` type, refer to `renv::settings$package.dependency.fields()` for the
+#'     `DESCRIPTION` fields included in the lockfile.
+#'     - **Custom files-based lockfile**: To specify custom files as the basis for the lockfile, set
+#'     `renv::settings$snapshot.type("custom")` and configure the `renv.snapshot.filter` option.
+#'
+#' @note
+#' This function computes the lockfile as a `promises::future_promise` promise, executing the process on a separate
+#' worker. If `future::plan()` is set to something other than `future::sequential`, it reuses the existing parallel
+#' backend. Otherwise, it sets `future::multisession` as the parallel plan with two workers. The `shiny::ExtendedTask()`
+#' is then used to run asynchronous computations. If the `teal` app started with the `future::sequential` plan, it is
+#' reset once the task is completed.
+#'
+#' @section lockfile usage:
+#' After creating the lockfile, you can restore the application environment using `renv::restore()`.
+#'
+#' @seealso [renv::snapshot()], [renv::restore()].
+#'
+#' @return Nothing. This function is executed for its side effect of creating a lockfile used in the `teal` application.
+#'
+#' @keywords internal
+teal_lockfile <- function() {
+  # If user has setup the file, there is no need to compute a new one.
+  user_lockfile <- getOption("teal.renv.lockfile", "")
+  if (!identical(user_lockfile, "")) {
+    if (file.exists(user_lockfile)) {
+      file.copy(user_lockfile, "teal_app.lock")
+      return(invisible(NULL))
+    } else {
+      stop("lockfile provided through options('teal.renv.lockfile') does not exist.")
+    }
+  }
+
+  if (!(is_in_test() || is_r_cmd_check())) {
+    old_plan <- future::plan()
+    # If there is already a parallel backend, reuse it.
+    if (inherits(old_plan, "sequential")) {
+      future::plan(future::multisession, workers = 2)
+    }
+
+    lockfile_task <- ExtendedTask$new(create_renv_lockfile)
```

Comment:
are you able to do `create_renv_lockfile("teal_app.lock")`
this way you will be having hardcoded filename inside one function only

Reply:
I did extend `create_renv_lockfile` with new parameter in here db2b95ccc8e8db3e9a89031cc5da2fdf99b92c76. You can pass it at `lockfile_task$invoke()`. This path is still hardcoded in `downloadHandler` and in tests.